### DISCUSSION TITLE
How to add a canonical tag into the <head> section?

### DISCUSSION THREAD

**Discussion by drganghe** [quarto-dev/quarto-cli#3976] [Q&A] [answered]
Google search seems to take website/ and website/index.html as two different pages and thinks one is a duplicate and therefore could not index it. And all page links in the sitemap file are with the  /index.html .

Add a <link> element with the attribute rel="canonical" to the <head> section of duplicate pages, pointing to the canonical page. 

Source: https://developers.google.com/search/docs/crawling-indexing/consolidate-duplicate-urls#rel-canonical-link-method

My question is how to do that in Quarto? Any thoughts will be appreciated.

**Comment by cderv** [comment]
@drganghe I believe Quarto website passed the canonical URL informations to Google seach using `sitemap.xml` (https://developers.google.com/search/docs/crawling-indexing/consolidate-duplicate-urls#sitemap-method) and help sync with Google by including it in robots.txt (https://developers.google.com/search/docs/crawling-indexing/sitemaps/build-sitemap#addsitemap). 

See for example in our Quarto website: 

* https://quarto.org/sitemap.xml
* https://quarto.org/robots.txt

Isn't that working on your website ? 

cc @dragonstyle as you may be interested here.

**Reply by drganghe** [reply]
Thank you for the detailed instruction. I have the sitemap setup and also added it to the robots.tex file. However, there are pages that still have indexing problems.

More general question: is there a way to add user-defined code or contents to the <head> section?

**Reply by cderv** [reply]
> However, there are pages that still have indexing problems.

Interesting. @dragonstyle we should probably look into this from our own quarto.org website to see if we can observe the same.

> More general question: is there a way to add user-defined code or contents to the section?

You can add content in HTML header using includes https://quarto.org/docs/output-formats/html-basics.html#includes
Maybe that is what you are looking for here to add a node in `<head>` ?

**Reply by drganghe** [reply]
Thank you Chris. I tried and the "includes" option helps to meet this need.

**Reply by dfolio** [reply]
As of today, I have the same issue with Google Search Console.
 It seems it find first  https://example.com/ (that are not in sitemap.xml) before looking for https://example.com/index.html (present in  sitemap.xml, but with no canonical) 

A simple hack to add a canonical (and to solve this issue) is to use metadata and the metadata.html [HTML Partials](https://quarto.org/docs/journals/templates.html)

## Example
Set in the yaml front mater of your document  `canonical: https://example.com/canonical.html`
and create/modify  metadata.html [HTML Partials](https://quarto.org/docs/journals/templates.html)  by adding
```html
$if(canonical)$<link rel="canonical" href="$canonical$" />$endif$
````

**Comment by mcb00** [comment]
I was also having similar issues with page indexing from Google Search Console. It seems like the fix is to add rel="canonical" tags to each canonical page's html header.

## Automatically add rel="canonical" link tags to pages
I wrote this python script [add-canonicals.py](https://github.com/mcb00/rr-blog/blob/2347dcaf34b3132566b6c2180f4f3f23fb0a65c6/add-canonicals.py) which reads page urls from the sitemap and then adds the correct canonical link tag in the html header of each page. You can automatically run the script after rendering by adding it to the root of your project and adding it as a post-render command in your `_quarto.yml` as

```
project:
  post-render:
    - "python add-canonicals.py"
```

**Comment by dfolio** [comment, marked as answer]
Since [quarto-1.4 release](https://quarto.org/docs/download/changelog/1.4/index.html#html-format) there is a support for [`canonical-url`](https://quarto.org/docs/reference/formats/html.html#canonical-url).
This is related to the [issues 4882](https://github.com/quarto-dev/quarto-cli/issues/4882), and can be considered solved.
